addon/PageScraper.js

```diff
@@ -0,0 +1,77 @@
+/* globals Services, Task */
+"use strict";
+
+const {MetadataParser} = require("addon/MetadataParser");
+const {Cu} = require("chrome");
+const options = require("@loader/options");
+
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.import("resource://gre/modules/Task.jsm");
+Cu.importGlobalProperties(["URL"]);
+
+const DEFAULT_OPTIONS = {
+  metadataTTL: 3 * 24 * 60 * 60 * 1000, // 3 days for the metadata to live
+  framescriptPath: new URL("data/page-scraper-content-script.js", options.prefixURI),
+  blackList: ["about:", "localhost", "resource://"]
+};
+
+function PageScraper(previewProvider, options = {}) {
+  this.options = Object.assign({}, DEFAULT_OPTIONS, options);
+  this._previewProvider = previewProvider;
+  this._metadataParser = new MetadataParser();
+}
+
+PageScraper.prototype = {
+  /**
+   * Check if the link is already in the metadata database. If not, parse the
+   *  HTML and insert it in the metadata database
+   */
+  _asyncParseAndSave: Task.async(function*(raw, url) {
+    let link = yield this._previewProvider.asyncDoesSingleLinkExist(url);
+    if (!link) {
+      let metadata = yield this._metadataParser.parseHTMLText(raw, url);
+      this._insertMetadata(metadata);
+    }
```

Comment:
We should definitely think of an update strategy, as of now it gets inserted in the metadata DB the same as any entry would (like from embedly), and is assigned a TTL of 3 days. After that there's auto expiry that happens in the DB and the next time they visit this page it won't be found in the DB and will get the metadata again for that page (acts as an update)